src/puppetlabs/trapperkeeper/services/status/status_core.clj

```diff
@@ -45,23 +45,40 @@
   (let [status-map (service-status-map svc-version status-version status-fn)]
     (swap! status-fns-atom update-in [svc-name] conj status-map)))
 
-(schema/defn ^:always-validate call-latest-status-fn-for-service :- ServiceStatus
+(schema/defn ^:always-validate call-status-fn-for-service :- ServiceStatus
   "Given a list of maps containing service information and a status function,
-  find the latest version, and return a map with the service's version, the
+  find the specified version, and return a map with the service's version, the
   version of the service's status, and the results of calling this status
-  function."
-  [service :- [ServiceInfo]
-   level :- ServiceStatusDetailLevel]
-  (let [latest-status (last (sort-by :service-status-version service))]
-    (assoc (select-keys latest-status [:service-version :service-status-version])
-           :status ((:status-fn latest-status) level))))
+  function. If no version for the status function is specified, the most recent
+  version will be used."
+  ([service-name :- schema/Str
+    service :- [ServiceInfo]
+    level :- ServiceStatusDetailLevel]
+    (call-status-fn-for-service service-name service level nil))
+  ([service-name :- schema/Str
+    service :- [ServiceInfo]
+    level :- ServiceStatusDetailLevel
+    service-status-version :- (schema/maybe schema/Int)]
+    (let [status (if (nil? service-status-version)
+                   (last (sort-by :service-status-version service))
+                   (first (filter #(= (:service-status-version %)
+                                      service-status-version)
+                                  service)))]
+      (when (nil? status)
+        (throw+ {:type    :service-status-version-not-found
+                 :message (str "No status function with version "
+                               service-status-version
```

Comment:
It will only be `nil` if a service version wasn't supplied. In that case, there must be at least one status function registered for the service, otherwise we'll return a `404` because we can't find anything for the service name. So I think this is fine.